docs/instrumentation_examples.md

```diff
@@ -0,0 +1,214 @@
+
+Recall the steps for a minimal working example (in [Measurement Overview](./metrics.md#instrumentation)):
+1. Create `Probe` via `get_probe(name)`
+2. Place `Probe` actions
+3. Create `Meter` with processing functions that take input from created `Probe`
+4. Connect `Meter` to `Hub` via `get_hub().connect_meter(meter)`
+
+The examples will show how each of these steps are accomplished.
+
+## Example 1: Measuring a Model Layer's Output
+### User Story
+I am interested in layer output from the second `relu` activation of a `forward` method located in `armory/baseline_models/pytorch/cifar.py`.
+### `Probe` Example Code
+The code below is an example of how to accomplish steps 1 and 2 (note the lines of code with `# added` comments at the end) for a model code that the user is modifying.
+```python
+"""
+CNN model for 32x32x3 image classification
+"""
+from typing import Optional
+
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+from art.estimators.classification import PyTorchClassifier
+
+from armory.instrument import get_probe # added
+probe = get_probe("my_model") # added
+
+DEVICE = torch.device("cuda" if torch.cuda.is_available() else "cpu")
+
+
+class Net(nn.Module):
+    """
+    This is a simple CNN for CIFAR-10 and does not achieve SotA performance
+    """
+
+    def __init__(self) -> None:
+        super(Net, self).__init__()
+        self.conv1 = nn.Conv2d(3, 4, 5, 1)
+        self.conv2 = nn.Conv2d(4, 10, 5, 1)
+        self.fc1 = nn.Linear(250, 100)
+        self.fc2 = nn.Linear(100, 10)
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        x = x.permute(0, 3, 1, 2)  # from NHWC to NCHW
+        x = self.conv1(x)
+        x = F.relu(x)
+        x = F.max_pool2d(x, 2)
+        x = self.conv2(x)
+        x = F.relu(x)
+        x_out = x.detach().cpu().numpy() # added
+        probe.update(layer_output=x_out) # added
+        x = F.max_pool2d(x, 2)
+        x = torch.flatten(x, 1)
+        x = self.fc1(x)
+        x = F.relu(x)
+        x = self.fc2(x)
+        output = F.log_softmax(x, dim=1)
+        return output
+
+...
+```
+
+
+
+#### Step 1
+After importing `get_probe` in line 11, `probe = get_probe("my_model")` in line 12 creates a `Probe` object with the name `"my_model"`, which is what the user can refer to later to apply processing functions through a `Meter` object.
+
+#### Step 2
+`x_out = x.detach().cpu().numpy()` in line 36 is taking the layer output of interest (second `relu` activation output) and converting the tensor to `numpy` array on the CPU, which will be passed to `probe`. An updated value of `x_out` is stored in `layer_output` via `probe.update(layer_output=x_out)` in line 37. Like the `Probe` name `"my_model"`, `layer_output` can be referenced by the user later to apply additional processing functions through a `Meter` object.
+
+### `Meter` Example Code
```

Comment:
my proposed rewriting of the text below `Meter Example Code`:

```
Now that a Probe instance has been created, we need to create a Meter object to accept any updated values from Probe and apply further processing that the user desires. We can create the Meter in a function added to a local Python script, then pass this function to the `user_init` field in our JSON config to ensure the code is run during scenario initialization, prior to scenario execution. Please refer to etc. etc. for more details on the user_init field and how to configure it.
```

Reply:
resolved with c24837a7